src/Bicep.Core/Emit/ForSyntaxValidatorVisitor.cs

```diff
@@ -247,6 +275,25 @@ private static bool IsPropertyLoop(SyntaxBase? parent, ForSyntax syntax)
             return parent is ObjectPropertySyntax property && ReferenceEquals(property.Value, syntax);
         }
 
+        /// <summary>
+        /// We cannot compile for-expressions when they are inside function calls or operators. This function
+        /// checks if the specified node allows for-expression usage.
+        /// </summary>
+        /// <param name="syntax">The node to check</param>
+        /// <returns></returns>
+        private static bool? IsPropertyLoopUsagePossibleInside(SyntaxBase syntax)
+        {
+            if(syntax is not ExpressionSyntax)
+            {
+                // not an expression but need to distinguish this state from "not allowed" in an expression node
+                return null;
+            }
+
+            // property loops can be used as long as the path to get to the property is constructed via object or array literals
+            // operators or function calls prevent usage of property loops inside
+            return syntax is ObjectSyntax or ObjectPropertySyntax or ArraySyntax or ArrayItemSyntax;
```

Comment:
IMO this function would be more readable with a switch statement - there are quite a few different syntax type checks happening on this line.

Reply:
Sure, can change it.

Reply:
Done